kernel/aster-nix/src/ipc/sem_set.rs

```diff
@@ -0,0 +1,180 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::{collections::btree_map::BTreeMap, vec::Vec};
+use core::time::Duration;
+
+use id_alloc::IdAlloc;
+use ostd::sync::{Mutex, RwMutex, RwMutexReadGuard, RwMutexWriteGuard};
+use spin::Once;
+
+use super::{key_t, IpcPermission};
+use crate::{
+    ipc::sem::Semaphore,
+    prelude::*,
+    process::{Gid, Uid},
+    time::SystemTime,
+};
+
+// The following constant values are derived from the default values in Linux.
+
+/// Maximum number of semaphore sets.
+pub const SEMMNI: usize = 32000;
+/// Maximum number of semaphores per semaphore ID.
+pub const SEMMSL: usize = 32000;
+/// Maximum number of seaphores in all semaphore sets.
+pub const SEMMNS: usize = SEMMNI * SEMMSL;
+/// Maximum number of operations for semop.
+pub const SEMOPM: usize = 500;
+/// MAximum semaphore value.
+pub const SEMVMX: i32 = 32767;
+/// Maximum value that can be recored for semaphore adjustment (SEM_UNDO).
+pub const SEMAEM: i32 = SEMVMX;
+
+static ID_ALLOCATOR: Once<Mutex<IdAlloc>> = Once::new();
+
+/// Semaphore sets in system
+static SEMAPHORE_SETS: RwMutex<BTreeMap<key_t, SemaphoreSet>> = RwMutex::new(BTreeMap::new());
+
+#[derive(Debug)]
+pub struct SemaphoreSet {
+    /// Number of semaphores in the set
+    nsems: usize,
+    /// Semaphores
+    sems: Vec<Semaphore>,
+    /// Semaphore permission
+    permission: IpcPermission,
+    /// Inner
+    inner: Mutex<Inner>,
+}
+
+#[derive(Debug)]
+struct Inner {
+    /// Last semop time
+    sem_otime: Duration,
+    /// Creation time or last modification via `semctl`
+    sem_ctime: Duration,
+}
+
+impl SemaphoreSet {
+    pub fn nsems(&self) -> usize {
+        self.nsems
+    }
+
+    pub fn get(&self, index: usize) -> Option<&Semaphore> {
+        self.sems.get(index)
+    }
+
+    pub fn permission(&self) -> &IpcPermission {
+        &self.permission
+    }
+
+    pub fn update_ctime(&self) {
+        let now = SystemTime::now()
+            .duration_since(&SystemTime::UNIX_EPOCH)
+            .unwrap();
+
+        self.inner.lock().sem_ctime = now;
+    }
+
+    pub fn update_otime(&self) {
+        let now = SystemTime::now()
+            .duration_since(&SystemTime::UNIX_EPOCH)
+            .unwrap();
+
+        self.inner.lock().sem_otime = now;
+    }
+
+    fn new(key: key_t, nsems: usize, mode: u16) -> Result<Self> {
+        debug_assert!(nsems <= SEMMSL);
+
+        let mut sems = Vec::with_capacity(nsems);
+        for _ in 0..nsems {
+            sems.push(Semaphore::new(0));
+        }
```

Comment:
The `vec![Semaphore::new(0); sems];` require the `Semaphore' to implement the clone trait, and I think the `Semaphore' should not implement the clone trait.

Reply:
Your words make sense to me! So it is ok to keep it as the original way.

But I see that the type of `sems` field becomes `sems: Box<[Arc<Semaphore>]>` in the latest commit, is it a mistake? Does the semaphore need to be cloned somewhere?

Reply:
Yes, this problem was discovered when gvisor testing was enabled. The problem is that using `sems: Box<[Semaphore]>` will cause the semaphore to hold the read lock of the `SEMAPHORE_SETS` when waiting for the semaphore to become available. However, there is a command in `semctl` to drop one of the semaphore_sets which requires the write lock of the `SEMAPHORE_SETS`.

Here I use `sems: Box<[Arc<Semaphore>]>` so that the semaphore will not hold the read lock.